File: src/main/java/cspfj/generator/ProblemGenerator.java
package cspfj.generator;

import java.util.Collection;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

import com.google.common.primitives.Ints;

import cspfj.exception.FailedGenerationException;
import cspfj.generator.constraint.GeneratorManager;
import cspfj.problem.BitVectorDomain;
import cspfj.problem.Domain;
import cspfj.problem.Problem;
import cspom.CSPOM;
import cspom.constraint.CSPOMConstraint;
import cspom.variable.BooleanDomain;
import cspom.variable.CSPOMDomain;
import cspom.variable.CSPOMVariable;

public final class ProblemGenerator {

    private ProblemGenerator() {
    }

    public static Problem generate(final CSPOM cspom)
            throws FailedGenerationException {

        // new ProblemCompiler(cspom).compile();

        final Problem problem = new Problem();

        for (CSPOMVariable v : cspom.getVariables()) {
            problem.addVariable(v.getName(), generateDomain(v.getDomain()));
        }

        final GeneratorManager gm = new GeneratorManager(problem);

        final Deque<CSPOMConstraint> queue = new LinkedList<CSPOMConstraint>(
                cspom.getConstraints());
        CSPOMConstraint firstFailed = null;
        while (!queue.isEmpty()) {
            final CSPOMConstraint constraint = queue.poll();

            if (gm.generate(constraint)) {
                firstFailed = null;
            } else if (firstFailed == constraint) {
                throw new FailedGenerationException(
                        "Could not generate the constraints "
                                + queue.toString());
            } else {
                if (firstFailed == null) {
                    firstFailed = constraint;
                }
                queue.offer(constraint);
            }
        }

        return problem;
    }

    private static Domain generateDomain(final CSPOMDomain<?> cspomDomain) {
        if (cspomDomain == null) {
            return null;
        }
        if (cspomDomain instanceof BooleanDomain) {
            final BooleanDomain bD = (BooleanDomain) cspomDomain;
            if (bD.isConstant()) {
                return new cspfj.problem.BooleanDomain(bD.getBoolean());
            }
            return new cspfj.problem.BooleanDomain();
        }
        return new BitVectorDomain(
                Ints.toArray((Collection<Integer>) cspomDomain.getValues()));
    }
}

File: src/main/scala/cspfj/generator/ProblemGenerator.scala
package cspfj.generator;

import cspfj.generator.constraint.GeneratorManager
import cspfj.{ Domain, IntDomain, Problem }
import cspom.constraint.CSPOMConstraint
import cspom.variable.{ CSPOMDomain, BooleanDomain }
import cspom.CSPOM
import scala.annotation.tailrec
import scala.collection.immutable.Queue
import cspom.variable.IntInterval
import cspfj.util.Loggable
import cspfj.UndefinedDomain

object ProblemGenerator extends Loggable {
  @throws(classOf[FailedGenerationException])
  def generate(cspom: CSPOM) = {

    // new ProblemCompiler(cspom).compile();

    val problem = new Problem();

    generateVariables(problem, cspom)

    val gm = new GeneratorManager(problem);

    var firstFailed: Option[CSPOMConstraint] = None;

    @tailrec
    def processQueue(queue: Queue[CSPOMConstraint]): Unit = if (queue.nonEmpty) {
      val (constraint, rest) = queue.dequeue

      if (gm.generate(constraint)) {
        firstFailed = None;
        processQueue(rest)
      } else firstFailed match {
        case Some(c) if c == constraint =>
          throw new FailedGenerationException(
            "Could not generate the constraints " + queue);
        case Some(_) =>
          processQueue(rest.enqueue(constraint));
        case None =>
          firstFailed = Some(constraint);
          processQueue(rest.enqueue(constraint));
      }

    }

    processQueue(Queue.empty ++ cspom.constraints)

    for (v <- problem.variables if v.constraints.isEmpty) {
      problem.removeVariable(v)
    }

    problem;
  }

  def generateVariables(problem: Problem, cspom: CSPOM) {
    for (v <- cspom.variables) {
      logger.fine("sizeD " + v.domainOption.map(_.size).getOrElse("?"))
      problem.addVariable(v.name, generateDomain(v.domainOption));
    }
  }

  def generateDomain[T](cspomDomain: Option[CSPOMDomain[T]]): Domain = cspomDomain map {
    case bD: BooleanDomain =>
      if (bD.isConstant) new cspfj.BooleanDomain(bD.getBoolean)
      else new cspfj.BooleanDomain();

    case int: IntInterval => IntDomain(int.lb to int.ub)

    case ext: CSPOMDomain[Int] => IntDomain(ext.values: _*)

    case _ => throw new UnsupportedOperationException
  } getOrElse UndefinedDomain
}
